src/index.js

```diff
@@ -126,169 +108,161 @@ var expressions = {
   /**
    */
 
-  $in: function(a, b) {
-    if (b instanceof Array) {
-      for (var i = b.length; i--; ) {
-        if (~a.indexOf(comparable(get(b, i)))) {
+  $in(a, b) {
+    if (Array.isArray(b)) {
+      const aSet = new Set(a);
+      for (let i = b.length; i--; ) {
+        if (aSet.has(comparable(get(b, i)))) {
           return true;
         }
       }
-    } else {
-      var comparableB = comparable(b);
-      if (comparableB === b && typeof b === "object") {
-        for (var i = a.length; i--; ) {
-          if (String(a[i]) === String(b) && String(b) !== "[object Object]") {
-            return true;
-          }
-        }
-      }
 
-      /*
-        Handles documents that are undefined, whilst also
-        having a 'null' element in the parameters to $in.
-      */
-      if (typeof comparableB == "undefined") {
-        for (var i = a.length; i--; ) {
-          if (a[i] == null) {
-            return true;
-          }
+      return false;
+    }
+
+    const comparableB = comparable(b);
+    if (comparableB === b && typeof b === 'object') {
+      for (let i = a.length; i--; ) {
+        if (String(a[i]) === String(b) && String(b) !== '[object Object]') {
+          return true;
         }
       }
+    }
 
-      /*
-        Handles the case of {'field': {$in: [/regexp1/, /regexp2/, ...]}}
-      */
-      for (var i = a.length; i--; ) {
-        var validator = createRootValidator(get(a, i), undefined);
-        var result = validate(validator, b, i, a);
-        if (
-          result &&
-          String(result) !== "[object Object]" &&
-          String(b) !== "[object Object]"
-        ) {
+    /*
+      Handles documents that are undefined, whilst also
+      having a 'null' element in the parameters to $in.
+    */
+    if (typeof comparableB === 'undefined') {
+      for (let i = a.length; i--; ) {
+        if (a[i] == null) {
           return true;
         }
       }
-
-      return !!~a.indexOf(comparableB);
     }
 
-    return false;
+    /*
+      Handles the case of {'field': {$in: [/regexp1/, /regexp2/, ...]}}
+    */
+    return maybeAsyncThen(
```

Comment:
So this is 'maybe run the internal filter as async and then return the result of `a.indexOf(comparableB) >= 0`'?

Reply:
Almost - it mirrors the prior code where it only considers `a.indexOf(comparableB) >= 0` if the possibly-async validators come up `false`.